cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -1407,7 +1407,7 @@ class AsofJoinNode : public ExecNode {
       ARROW_ASSIGN_OR_RAISE(
           auto input_state,
           InputState::Make(i, tolerance_, must_hash_, may_rehash_, key_hashers_[i].get(),
-                           this, inputs[i], backpressure_counter_,
+                           inputs[i], this, backpressure_counter_,
```

Comment:
Simply reverting this line where the bug is introduced:
https://github.com/apache/arrow/pull/34392/files#diff-5493b6ae7ea2a4d5cfb581034c076e9c4be7608382168de6d1301ef67b6c01eeR1410

@rtpsw is working on adding tests but it could take some time.

Reply:
Note that I have just added tests in #35874, though this PR can proceed, if so desired.